client/services/ReinforcementGeneratorService.js

```diff
@@ -0,0 +1,101 @@
+// Copyright 2017 The TIE Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service for generating positive reinforcement when a user
+ * submits code based on the tasks that the code passes/fails on.
+ */
+
+tie.factory('ReinforcementGeneratorService', [
+  'ReinforcementObjectFactory', 'TranscriptService',
+  function(ReinforcementObjectFactory, TranscriptService) {
+    return {
+      getReinforcement: function(task, codeEvalResult) {
+
+        var testsKeyedByTag = function(tests) {
+          var splitTests = {};
+          for (var i = 0; i < tests.length; ++i) {
+            var test = tests[i];
+            test.originalIndex = i;
+            var testTag = test.getTag();
+            if (splitTests.hasOwnProperty(testTag)) {
+              splitTests[testTag].push(test);
+            } else {
+              splitTests[testTag] = [test];
+            }
+          }
+          return splitTests;
+        };
+
+        var previousSnapshot = TranscriptService.getPreviousSnapshot();
+        var previousReinforcement = null;
+        if (previousSnapshot !== null) {
+          previousReinforcement = previousSnapshot.getReinforcement();
+        }
+
+        var reinforcement = ReinforcementObjectFactory.create(task);
+
+        // Copy the previous reinforcement object if we are on the same task.
+        if (previousReinforcement !== null &&
+            task === previousReinforcement.getTask()) {
+          for (var tag in previousReinforcement.getPassedList()) {
```

Comment:
Define a local variable for previousReinforcement.getPassedList() and use that instead; don't keep calling the method repeatedly (especially in a for loop). Ditto for the past fails list.

It may make sense to add a reinforcement.clonePassedList() method that uses angular.copy() to bulk-populate the list; if you go that route, make sure to document the expected inputs.